### PR TITLE
Allow serviceAccount to reuse auth tokens.

### PR DESCRIPTION
**Pull request by jmuk**
jwt.Config.TokenSource() actually returns a reusable token source,
but serviceAccount actually discards the reusability, it always
creates a new token source instance.

This patch could introduces a redundancy when the token expires;
s.tokenSource.Token() implicitly attempts to obtain the new token
with the old context (the context which s.tokenSource was created
with), and if it fails, a new token source will be created with
a new context and then it attempts to obtain the token again.

I believe this redundancy is ignorable mostly, because this method
falls into one of the following scenarios:
- the same context is used among a program or the current sequence
  of gRPC calls. Then the 'old context' and the 'new context' are
  actually same.
- the old context has enough time by the deadline; simply
  obtaining the new token with the old context succeeds. The new
  context is not used at all.
- the old context is already canceled (or deadline exceeded). Then
  the attempt with the old context fails immediately and switches
  to create a new context.

On the other hand, cancelling the old context or some deadline
settings could lead to the inefficient code path theoretically.
I think that is quite rare though.

### CODE REVIEW

**Comment by iamqizhao** [comment]
@adg Andrew, can you help review this?

File: credentials/oauth/oauth.go [hunk 1]
```diff
@@ -142,7 +157,7 @@ func (s serviceAccount) GetRequestMetadata(ctx context.Context, uri ...string) (
 	}, nil
 }
 
-func (s serviceAccount) RequireTransportSecurity() bool {
+func (s *serviceAccount) RequireTransportSecurity() bool {
```
Comment (adg) [inline credentials/oauth/oauth.go]: Why change this?

**Comment by adg** [comment]
If you want to re-use tokens then you should cache the token and fetch a new one when it has expired. There's no need to re-use the token source; the setup for that is just allocating a couple of structs.

**Comment by iamqizhao** [comment]
any updates?

**Comment by jmuk** [comment]
@adg 
I'm not sure what's your point.

To me, 
```
serviceAccount, _ := oauth.NewServiceAccountFromFile(filename, scopes...)
conn, _ := grpc.Dial(service, grpc.WithPerRPCCredentials(serviceAccount))
```
it's not so obvious this code doesn't reuse the token.

And this happens not because serviceAccount can't create reusable token source, but because serviceAccount creates a reusable token source and then discard it every time a token is required. And, I think it's not great to put the logic of reusing token (and refreshing it when necessary) here, it already exists in golang.org/x/oauth2 as reuseTokenSource; and this patch bring reuseTokenSource here.

File: credentials/oauth/oauth.go [hunk 2]
```diff
@@ -129,11 +129,26 @@ func NewComputeEngine() credentials.Credentials {
...
+	config      *jwt.Config
+	tokenSource oauth2.TokenSource
+}
+
+func (s *serviceAccount) GetRequestMetadata(ctx context.Context, uri ...string) (map[string]string, error) {
+	var token *oauth2.Token
+	var err error
+	if s.tokenSource != nil {
+		token, err = s.tokenSource.Token()
+		if err != nil {
```
Comment (adg) [inline credentials/oauth/oauth.go]: This if block should be redundant. Token() shouldn't return a non-nil token with a non-nil error.

File: credentials/oauth/oauth.go [hunk 3]
```diff
@@ -129,11 +129,26 @@ func NewComputeEngine() credentials.Credentials {
...
+			token = nil
+		}
+	}
+	if token == nil {
```
Comment (adg) [inline credentials/oauth/oauth.go]:
```
if token == nil || err != nil {
```

seems clearer.

File: credentials/oauth/oauth.go [hunk 4]
```diff
@@ -129,11 +129,26 @@ func NewComputeEngine() credentials.Credentials {
...
+		tokenSource := s.config.TokenSource(ctx)
```
Comment (adg) [inline credentials/oauth/oauth.go]: `ts` is a fine variable name for a tokensource

File: credentials/oauth/oauth.go (hunk 2 above)
Comment (adg) [inline credentials/oauth/oauth.go]: if err == nil {
  return map[string]string{"authorization": token.TokenType + " " + token.AccessToken}
}

File: credentials/oauth/oauth.go (hunk 3 above)
Comment (adg) [inline credentials/oauth/oauth.go]: drop this if block

File: credentials/oauth/oauth.go [hunk 5]
```diff
@@ -129,11 +129,26 @@ func NewComputeEngine() credentials.Credentials {
...
+		token, err = tokenSource.Token()
+		if err == nil {
```
Comment (adg) [inline credentials/oauth/oauth.go]:
```
if err != nil {
  return nil, err
}
s.tokenSource = ts
```

File: credentials/oauth/oauth.go [hunk 6]
```diff
@@ -129,11 +129,26 @@ func NewComputeEngine() credentials.Credentials {
...
+			s.tokenSource = tokenSource
+		}
+	}
 	if err != nil {
```
Comment (adg) [inline credentials/oauth/oauth.go]: delete

**Comment by adg** [comment]
OK fair enough. I left some comments on how the code could be structured a little more idiomatically.

**Comment by dfawley** [comment]
This approach does not work, as I found in PR #1238.  I will fix this with my updates to that PR.

**Pull request closed without merging**